GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    ///   The GVFS NamePipeProtocol:
+    ///       GVFS can send "messages" over the wire. A message is defined as a
+    ///       sequence of UTF8 encoded text whose end is delineated by a TerminatorByte.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
+
+        private byte[] buffer;
+        private int bufferContentLength;
+
+        private Stream stream;
+
+        public NamedPipeStreamReader(Stream stream)
+        {
+            this.stream = stream;
+            this.buffer = new byte[NamedPipeStreamReader.BufferLength];
+        }
+
+        public string ReadMessage()
+        {
+            List<byte> bytes = new List<byte>();
+
+            int bytesRead = this.Read();
+            if (bytesRead == 0)
+            {
+                return null;
+            }
+
+            while (true)
+            {
+                if (this.buffer[bytesRead - 1] == TerminatorByte)
+                {
+                    bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, bytesRead - 1));
+                    break;
+                }
+
+                bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, bytesRead));

```

Comment:
What are the perf implications, if any, of using an `ArraySegment` here?

Reply:
`ArraySegment` is a struct that contains a reference to the underlying array and an offset into it - It is not a full copy of the array.

`ArraySegment` also Implements `ICollection`, which means `AddRange` can copy the items using `CopyTo` and `Array.Copy`, instead of inserting items individually (If I am reading the underlying source correctly).